### PR TITLE
Implement `lists::reverse`

### PR DESCRIPTION
**Pull request by ttnghia**
This implements `lists::reverse` that output a lists column in which each list is generated by reversing the order of the elements in the corresponding input list.

Example:
```
 s = [ [1, 2, 3], [], null, [4, 5, null] ]
 r = reverse(s)
 r is now [ [3, 2, 1], [], null, [null, 5, 4] ]
```

This is to support https://github.com/NVIDIA/spark-rapids/issues/4375 and https://github.com/NVIDIA/spark-rapids/issues/6885.

### CODE REVIEW

File: cpp/src/lists/reverse.cu
Comment (PointKernel) [inline cpp/src/lists/reverse.cu:39]: It would be great to declare this function in `include/cudf/lists/detail/reverse.hpp`. Otherwise, this can be scoped in an anonymous namespace

File: cpp/src/lists/utilities.cu
Comment (PointKernel) [inline cpp/src/lists/utilities.cu:61]:
```suggestion
  if (input.is_empty()) { return std::make_unique<column>(input.offsets(), stream, mr); }
```

**Review by Ethyling — APPROVED** [review]

**Review by vyasr — APPROVED** [review]
CMake LGTM.

**Review by hyperbolic2346 — APPROVED** [review]

**Comment by bdice** [comment]
This PR is waiting on a review from @rapidsai/cudf-java-codeowners.

**Review by bdice — APPROVED** [review]

**Review by PointKernel — APPROVED** [review]

**Review by jlowe — APPROVED** [review]

**Comment by ttnghia** [comment]
@gpucibot merge

**Pull request merged**
